src/main/java/appeng/helpers/IPatternTerminalHost.java

```diff
@@ -0,0 +1,54 @@
+package appeng.helpers;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+import appeng.api.inventories.InternalInventory;
+import appeng.api.networking.security.IActionHost;
+import appeng.api.storage.ITerminalHost;
+import appeng.util.inv.InternalInventoryHost;
+
+public interface IPatternTerminalHost extends ITerminalHost, IActionHost, InternalInventoryHost {
+
+    /**
+     * Identifies the sub-inventory used by the pattern terminal to encode the inputs of crafting or processing
+     * patterns.
+     */
+    ResourceLocation INV_CRAFTING = new ResourceLocation(
+            "appliedenergistics2:pattern_terminal_crafting");
+
+    /**
+     * Identifies the sub-inventory used by the pattern terminal to encode the outputs of crafting or processing
+     * patterns.
+     */
+    ResourceLocation INV_OUTPUT = new ResourceLocation(
+            "appliedenergistics2:pattern_terminal_output");
+
+    @Nullable
+    InternalInventory getSubInventory(ResourceLocation id);
+
+    boolean isCraftingRecipe();
+
+    void setCraftingRecipe(final boolean craftingMode);
+
+    boolean isSubstitution();
+
+    void setSubstitution(final boolean canSubstitute);
+
+    boolean isFluidSubstitution();
+
+    void setFluidSubstitution(boolean canSubstitute);
+
+    default void fixCraftingRecipes() {
```

Comment:
What's this for?

Reply:
It came from PatternTerminalPart

I tried moving as much methods as possible from PatternTerminalPart to IPatternTerminalHost
turns out not much was possible